replayserver/streams/base.py

```diff
@@ -154,22 +173,76 @@ class ConcreteDataMixin:
     def __init__(self):
         self._header = None
         self._data = bytearray()
+        self._discarded_data = 0
+        self._len = 0
+
+    def _add_data(self, data):
+        if self._discarded_data <= self._len:
+            self._data += data
+        else:
+            self._data += data[self._discarded_data - self._len:]
+        self._len += len(data)
 
     @property
     def header(self):
         return self._header
 
     def _data_length(self):
-        return len(self._data)
+        return self._len
 
     def _data_slice(self, s):
+        if isinstance(s, slice):
+            return self._get_slice(s)
+        if s >= 0:
+            if s < self._discarded_data:
+                raise IndexError
+            s -= self._discarded_data
         return self._data[s]
 
+    def _get_slice(self, s):
+        s, e, st = s.start, s.stop, s.step
```

Comment:
`M != a`

`!` = do not
`M` = I
`a` = anagram
`=` = like

Reply:
not going to use verbose names for mucking around with arithmetics, shorthands are obvious

Reply:
There's actually an `indices` method of `slice` that avoids this mucking about. Good to know, glad this code was revisited.